### PR TITLE
[Feature](Nereids)add RelationId as a unique identifier of relations

### PR DESCRIPTION
**Pull request by sohardforaname**
# Proposed changes

Issue Number: close #xxx

## Problem summary

In Nereids, we could not distinguish two relation from same table in one PlanTree.
This lead to some trick code to process them when do plan. Such as a separate branch to do equals in GroupExpression.
This PR add RelationId to LogicalRelation and PhysicalRelation. Then all relations equals function will compare RelationId to help us distinguish two relation from same table.

TODO:
add relation id to UnboundRelation, UnboundOneRowRelation, LogicalOneRowRelation, PhysicalOneRowRelation.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java [hunk 1]
```diff
@@ -47,6 +51,10 @@ public StatementBase getParsedStatement() {
         return parsedStatement;
     }
 
+    public RelationId getRelationId() {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java:54]:
```suggestion
    public RelationId nextRelationId() {
```

**Comment by wangshuo128** [comment]
In general, I have a question.
Why do we generate a relation id when binding relation instead of generation a unique relation id when creating `UnboundRelation` in the parser?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java [hunk 2]
```diff
@@ -81,7 +86,8 @@ public boolean equals(Object o) {
             return false;
         }
         LogicalRelation that = (LogicalRelation) o;
-        return Objects.equals(table.getId(), that.table.getId()) && Objects.equals(qualifier, that.qualifier);
+        return Objects.equals(id, that.id)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java:89]: just check relation id in hashcode and equals

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -165,6 +167,8 @@
  */
 public class LogicalPlanBuilder extends DorisParserBaseVisitor<Object> {
 
+    private final IdGenerator<RelationId> idGenerator = RelationId.createGenerator();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:170]: add a todo, we need to move it to statement context finally

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java [hunk 4]
```diff
@@ -41,35 +42,40 @@
  * Represent a relation plan node that has not been bound.
  */
 public class UnboundRelation extends LogicalLeaf implements Relation, Unbound {
+
+    public final RelationId id;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:46]: private, and add a getter

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/RelationId.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    public static IdGenerator<RelationId> createGenerator() {
+        return new IdGenerator<RelationId>() {
+            @Override
+            public RelationId getNextId() {
+                return new RelationId(nextId++);
+            }
+
+            @Override
+            public RelationId getMaxId() {
+                return new RelationId(nextId++);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/RelationId.java:57]:
```suggestion
                return new RelationId(nextId);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java [hunk 6]
```diff
@@ -107,7 +109,8 @@ public boolean equals(Object o) {
             return false;
         }
         PhysicalOlapScan that = (PhysicalOlapScan) o;
-        return selectedIndexId == that.selectedIndexId
+        return Objects.equals(id, that.id)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java:112]: just compare id

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRelation.java [hunk 7]
```diff
@@ -36,24 +37,29 @@
  */
 public abstract class PhysicalRelation extends PhysicalLeaf implements Scan {
 
+    public final RelationId id;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRelation.java:40]: private

**Review by morrySnow — COMMENTED** [review]
u need to change group expression's equals to use relation id to compare relations. and remove bind one row relation and bind empty set's TODO

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java [hunk 8]
```diff
@@ -125,4 +132,7 @@ public List<Long> getSelectedPartitionIds() {
         return selectedPartitionIds;
     }
 
+    public RelationId getId() {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java:135]: If this method is common for both logical and physical scan, you'd better to an declaration in the `Scan` interface

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java [hunk 9]
```diff
@@ -44,8 +45,10 @@ public abstract class LogicalRelation extends LogicalLeaf implements Scan {
     protected final List<String> qualifier;
     protected final List<Long> selectedPartitionIds;
 
-    public LogicalRelation(PlanType type, Table table, List<String> qualifier) {
-        this(type, table, qualifier, Optional.empty(), Optional.empty(), Collections.emptyList());
+    private final RelationId id;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java:48]:
```suggestion
    protected final RelationId id;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRelation.java [hunk 10]
```diff
@@ -38,22 +39,27 @@ public abstract class PhysicalRelation extends PhysicalLeaf implements Scan {
 
     protected final List<String> qualifier;
 
+    private final RelationId id;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRelation.java:42]:
```suggestion
    protected final RelationId id;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java [hunk 11]
```diff
@@ -103,14 +105,7 @@ public boolean equals(Object o) {
...
         }
-        if (o == null || getClass() != o.getClass() || !super.equals(o)) {
-            return false;
-        }
-        PhysicalOlapScan that = (PhysicalOlapScan) o;
-        return selectedIndexId == that.selectedIndexId
-                && Objects.equals(selectedTabletIds, that.selectedTabletIds)
-                && Objects.equals(selectedPartitionIds, that.selectedPartitionIds)
-                && Objects.equals(olapTable, that.olapTable);
+        return o != null && getClass() == o.getClass() && super.equals(o);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java:108]: why change this, i think we decide equals and hashCode should contains all attributes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 12]
```diff
@@ -107,28 +110,29 @@ public boolean equals(Object o) {
         if (o == null || getClass() != o.getClass() || !super.equals(o)) {
             return false;
         }
-        return true;
+        return Objects.equals(selectedPartitionIds, ((LogicalOlapScan) o).selectedPartitionIds)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:113]: need override hashCode function

**Review by morrySnow — APPROVED** [review]

**Pull request merged**